### ISSUE TITLE
need function-level reset support

### ISSUE THREAD

**Issue by jlevon** [nutanix/libvfio-user#461]
Certain writes to the PCI Express capability are supposed to initiate a function-level reset (FLR). This should only happen if pxcap::pxdcap::flrc is set.

NVMe controllers are required to support FLR, and nvmf is dutifully setting this bit in the pxcap. But we just ignore writes to pxcap::pxdc::iflr today.

We need to add a new reset callback type to support FLR, wire it up in SPDK, and reset at least some of our internal state (not clear how much, but certainly not including detaching the socket client).

Cc @changpe1

**Comment by changpe1** [comment]
OK, SPDK nvmf can support controller level reset, FLR will cause a controller level reset, some internal states are required to clear, such as live migration state.

**Comment by tmakatos** [comment]
@swapnili can we close this?

**Comment by jlevon** [comment]
we can either keep this open to track the SPDK-side work, or open a new ticket

**Comment by swapnili** [comment]
Anything is fine, I kept it open just to track SPDK work @changpe1 is working on.

**Comment by changpe1** [comment]
> Anything is fine, I kept it open just to track SPDK work @changpe1 is working on.

yeah, I'm working on it, let's keep it open until SPDK support it.

**Comment by tmakatos** [comment]
From the spec (NVMe® over PCIe® Transport Specification, Revision 1.2), 3.8.5.3:

> Function Level Reset Capability (FLRC): A value of ‘1’ indicates the Function
supports the optional Function Level Reset mechanism. NVM Express controllers shall
support Function Level Reset.

**Comment by jlevon** [comment]
We've done the necessary plumbing in libvfio-user, seems silly to have this open for SPDK changes. So closing in favour of https://github.com/spdk/spdk/issues/3903 need proper FLR support in vfio_user.c
